dotCMS/src/main/java/com/dotcms/storage/ContentletMetadataAPIImpl.java

```diff
@@ -0,0 +1,255 @@
+package com.dotcms.storage;
+
+import com.dotcms.contenttype.model.field.BinaryField;
+import com.dotcms.contenttype.model.field.Field;
+import com.dotcms.contenttype.model.field.FieldVariable;
+import com.dotmarketing.business.APILocator;
+import com.dotmarketing.business.CacheLocator;
+import com.dotmarketing.business.FactoryLocator;
+import com.dotmarketing.portlets.contentlet.business.ContentletCache;
+import com.dotmarketing.portlets.contentlet.model.Contentlet;
+import com.dotmarketing.util.Config;
+import com.dotmarketing.util.Logger;
+import com.dotmarketing.util.StringUtils;
+import com.dotmarketing.util.UtilMethods;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.liferay.util.StringPool;
+import io.vavr.Tuple;
+import io.vavr.Tuple2;
+import io.vavr.control.Try;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Default implementation
+ * @author jsanca
+ */
+public class ContentletMetadataAPIImpl implements ContentletMetadataAPI {
+
+    private final FileStorageAPI fileStorageAPI;
+
+    public ContentletMetadataAPIImpl() {
+        this(APILocator.getFileStorageAPI());
+    }
+
+    @VisibleForTesting
+    public ContentletMetadataAPIImpl(final FileStorageAPI fileStorageAPI) {
+        this.fileStorageAPI = fileStorageAPI;
+    }
+
+    @Override
+    public ContentletMetadata generateContentletMetadata(final Contentlet contentlet,
+                                                         final Set<String> basicBinaryFieldNameSet,
+                                                         final Set<String> fullBinaryFieldNameSet) throws IOException {
+
+        final ImmutableMap.Builder<String, Map<String, Object>> fullMetadataMap  = new ImmutableMap.Builder();
+        final ImmutableMap.Builder<String, Map<String, Object>> basicMetadataMap = new ImmutableMap.Builder();
+        final  Map<String, com.dotcms.contenttype.model.field.Field>  fieldMap   = contentlet.getContentType().fieldMap();
+        /*
+		Verify if it is enabled the option to always regenerate metadata files on reindex,
+		enabling this could affect greatly the performance of a reindex process.
+		 */
+        final boolean alwaysRegenerateMetadata = Config
+                .getBooleanProperty("always.regenerate.metadata.on.reindex", false);
+        Map<String, Object> metadataMap        = null;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid declaring a variable if it is unreferenced before a possible exit point.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5547873)